prompting/tools/datasets/wiki.py

```diff
@@ -0,0 +1,270 @@
+            selector (Selector, optional): _description_. Defaults to None.
+            include (List, optional): _description_. Defaults to None.
+            exclude (List, optional): _description_. Defaults to None.
+
+        Returns:
+            Dict: _description_
+        """
+
+        page = _get_page(title=name, **kwargs)
+        if page is None:
+            return None
+
+        # Only return a sections with a minimum number of words
+        exclude = (exclude or []) + list(self.EXCLUDE_HEADERS)
+        sections = process_page(page,
+                                valid_header=lambda x: x not in exclude and (not include or x in include),
+                                valid_content=lambda x: len(x.split())>=self.min_length_words
+                                )
+
+        key = header, section_title = selector(list(sections.keys()))
+        content = '\n'.join(sections[key])
+        section_length = len(content.split())
+        return {
+            "title": name, # title of wiki article
+            "topic": header or section_title, # title of wiki section
+            'subtopic': section_title,
+            'content': content,
+            'internal_links': list(filter(lambda x: x not in exclude, page.sections)),
+            'external_links': most_relevant_links(page, num_links=self.max_links), 
+            'tags': filter_categories(page.categories, exclude=self.EXCLUDE_CATEGORIES),
+            'source': 'Wikipedia',
+            'extra': {'url': page.url, 'page_length': len(page.content.split()), 'section_length': section_length},
+        }
+
+    def search(self, name, results=3, selector: Selector = None) -> Dict:
+        titles = _wiki_search(name, results=results)
+        title = selector(titles)
+        return self.get(title, selector=selector)
+
+    def random(self, pages=10, seed=None, selector: Selector = None, **kwargs) -> Dict:
+        titles = wiki.random(pages=pages) if seed is None else _get_random_titles(pages=pages, seed=seed)
+        title = selector(titles)
+        return self.get(title, selector=selector)
+
+
+
+
+class WikiDateDataset(Dataset):
+
+    INCLUDE_HEADERS = ("Events", "Births", "Deaths")
+    MONTHS = ("January", "February", "March", "April", "May", "June", "July", "August", "September", "October", "November", "December")
+    EXCLUDE_CATEGORIES = ('articles', 'wiki', 'pages', 'cs1')
+
+    def __init__(self, max_tries: int = 10, seed=None):
+        self.max_tries = max_tries
+        self.seed = seed
+        self.rng = random.Random(seed)
+
+    def _random_date(self, year: int = None, month: int = None) -> int:
+        """Returns a random date in the format "Month_DD" (e.g., "January_01")."""
+        if year is None:
+            year = self.rng.randint(0, 2024)
+        if month is None:
+            month = self.rng.randint(1, 12)
+
+        max_days = 31 if month in (1, 3, 5, 7, 8, 10, 12) else 30
+        max_days = max_days if month != 2 else 29
```

Comment:
It seems that we are not handling possible leap year scenarios, which could break the code with `ValueError: day is out of range for month`, even though its probability of happening is small